From de11d3a28cbdc680eba6ba08d7e39e7ee3ea5039 Mon Sep 17 00:00:00 2001
From: Yunfeng Ye <yeyunfeng@huawei.com>
Date: Wed, 29 Dec 2021 12:13:56 +0800
Subject: [PATCH] arm64: mm: support setting page attributes for debugging
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: de11d3a28cbdc680eba6ba08d7e39e7ee3ea5039


euleros inclusion
category: feature
feature: Memory debug feature
bugzilla: https://gitee.com/openeuler/kernel/issues/I4MTG7

-------------------------------------------------

When pagealloc debug is enabled, block mappings or contiguous hints are
no longer used for linear address area. Therefore, support setting page
attributes in this case is useful for debugging memory corruption
problems.

Signed-off-by: Yunfeng Ye <yeyunfeng@huawei.com>
Reviewed-by: Kefeng Wang <wangkefeng.wang@huawei.com>
Signed-off-by: Yang Yingliang <yangyingliang@huawei.com>
Reviewed-by: Chao Liu <liuchao173@huawei.com>
Reviewed-by: Kefeng Wang <wangkefeng.wang@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 arch/arm64/mm/pageattr.c | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/arch/arm64/mm/pageattr.c b/arch/arm64/mm/pageattr.c
index 1b94f5b82654..0bc12dbf2843 100644
--- a/arch/arm64/mm/pageattr.c
+++ b/arch/arm64/mm/pageattr.c
@@ -81,8 +81,16 @@ static int change_memory_common(unsigned long addr, int numpages,
 	area = find_vm_area((void *)addr);
 	if (!area ||
 	    end > (unsigned long)area->addr + area->size ||
-	    !(area->flags & VM_ALLOC))
+	    !(area->flags & VM_ALLOC)) {
+		/*
+		 * When pagealloc debug is enabled, the linear address is
+		 * mapped with NO_BLOCK_MAPPINGS and NO_CONT_MAPPINGS flags.
+		 */
+		if (numpages && debug_pagealloc_enabled())
+			return __change_memory_common(start, size,
+				set_mask, clear_mask);
 		return -EINVAL;
+	}
 
 	if (!numpages)
 		return 0;
-- 
2.26.2

